### PR TITLE
 Warns users if they put a URL into allowed_domains (#2250)

### PR DESCRIPTION
**Pull request by jlong49**
Warns users if they put a URL into allowed_domains and corrects that URL to an actual domain via urlparse.  (#2250)

### CODE REVIEW

**Comment by jlong49** [comment]
This build passed all of the tests in Tox but obviously failed the Travis tests. If anyone gets a chance, would they mind explaining to me why this happened and what I did incorrectly?

**Comment by eLRuLL** [comment]
Also, please correct title and commit message, it is `allowed` not `alowed` domains. I would also recommend shortening the commit message.

**Comment by redapple** [comment]
hi @jlong49 ,

> This build passed all of the tests in Tox but obviously failed the Travis tests.

When I pull your branch and run `tox -e py27`, I do not see that all tests pass:

```
(stripped)
_____________________________________________________________________________________________ TestOffsiteMiddleware4.test_process_spider_output _____________________________________________________________________________________________

self = <tests.test_spidermiddleware_offsite.TestOffsiteMiddleware4 testMethod=test_process_spider_output>

    def setUp(self):
        crawler = get_crawler(Spider)
        self.spider = crawler._create_spider(**self._get_spiderargs())
        self.mw = OffsiteMiddleware.from_crawler(crawler)
>       self.mw.spider_opened(self.spider)

/home/paul/src/scrapy/tests/test_spidermiddleware_offsite.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/paul/src/scrapy/scrapy/spidermiddlewares/offsite.py:72: in spider_opened
    self.host_regex = self.get_host_regex(spider)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <scrapy.spidermiddlewares.offsite.OffsiteMiddleware object at 0x7f3e4d5034d0>, spider = <Spider 'foo' at 0x7f3e4deab3d0>

    def get_host_regex(self, spider):
        """Override this method to implement a different offsite policy"""
        allowed_domains = getattr(spider, 'allowed_domains', None)
        try:
            from urllib import urlparse
        except ImportError:
>           from urllib.parse import urlparse
E           ImportError: No module named parse

/home/paul/src/scrapy/scrapy/spidermiddlewares/offsite.py:58: ImportError
----------------------------------------------------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------------------------------------------------
2017-01-23 11:50:49 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.logstats.LogStats',
 'scrapy.extensions.telnet.TelnetConsole']
2017-01-23 11:50:49 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.logstats.LogStats',
 'scrapy.extensions.telnet.TelnetConsole']
================================================================================= 13 failed, 1463 passed, 5 skipped, 13 xfailed, 2 error in 257.15 seconds ==================================================================================
ERROR: InvocationError: '/home/paul/src/scrapy/.tox/py27/bin/py.test --cov=scrapy --cov-report= scrapy tests'
__________________________________________________________________________________________________________________ summary __________________________________________________________________________________________________________________
ERROR:   py27: commands failed
```

It's the same as on Travis for Python 2.7: see https://travis-ci.org/scrapy/scrapy/jobs/192922050

So there's (at least) a problem with how `urlparse` is imported in Python 2.7.
If I change your patch with the following:

```
$ git diff
diff --git a/scrapy/spidermiddlewares/offsite.py b/scrapy/spidermiddlewares/offsite.py
index 62c1992..45412fe 100644
--- a/scrapy/spidermiddlewares/offsite.py
+++ b/scrapy/spidermiddlewares/offsite.py
@@ -53,7 +53,7 @@ class OffsiteMiddleware(object):
         """Override this method to implement a different offsite policy"""
         allowed_domains = getattr(spider, 'allowed_domains', None)
         try:
-            from urllib import urlparse
+            from urlparse import urlparse
         except ImportError:
             from urllib.parse import urlparse
         if not allowed_domains:
```

I am able to run `tox -e py27` with all passing tests.
I'll leave you run and fix all tests locally and tests on Travis too.

Also note that for `urlparse`, scrapy uses `from six.moves.urllib.parse import urlparse` in [many](https://github.com/scrapy/scrapy/blob/129421c7e31b89b9b0f9c5f7d8ae59e47df36091/scrapy/utils/url.py#L10) [places](https://github.com/scrapy/scrapy/blob/129421c7e31b89b9b0f9c5f7d8ae59e47df36091/scrapy/utils/httpobj.py#L5) (if not everywhere where it's needed). I suggest you do the same.

File: scrapy/spidermiddlewares/offsite.py [hunk 1]
```diff
@@ -52,6 +53,13 @@ def get_host_regex(self, spider):
         allowed_domains = getattr(spider, 'allowed_domains', None)
         if not allowed_domains:
             return re.compile('') # allow all by default
+        else:
+            for i, user_domain in enumerate(allowed_domains):
+                parsed_uri = urlparse(user_domain)
+                if user_domain != parsed_uri.netloc and parsed_uri.netloc != '':
+                    allowed_domains[i] = parsed_uri.netloc
+                    logging.warn("%s is an invalid domain and has automatically been converted to %s" % (
```
Comment (kmike) [inline scrapy/spidermiddlewares/offsite.py]: it should be logger.warn, so that it is clear where the message comes from

**Comment by kmike** [comment]
Hey @jlong49! Thanks for the PR. Could you please add a [test](https://github.com/scrapy/scrapy/blob/master/tests/test_spidermiddleware_offsite.py) for this feature?

**Comment by jlong49** [comment]
@kmike yes I will in the next few days, sorry for the delay in replying

**Comment by qhuang872** [comment]
Hi everyone, I just tried to added a note on the docs, to remind user to add domain name into the list instead of the whole url, I did not modify any other code at all, and my pull request did not pass the travis test, can anyone advise me why is that? I saw jlong49 has similar issue, but I wasn't implementing the 'auto correct' feature, just editing the doc.

**Comment by redapple** [comment]
@qhuang872 , [the failed Travis build](https://travis-ci.org/scrapy/scrapy/builds/213244076) for [your PR](https://github.com/scrapy/scrapy/pull/2670) has some explanation on what was wrong
```
Warning, treated as error:
/home/travis/build/scrapy/scrapy/docs/topics/spiders.rst:84: ERROR: Unexpected indentation.
```
It's a restructured text syntax error and I see that you have removed the note to make the build pass. I think `.. note::` should be seperated from the previous paragraph by a newline. This is what is done for the point just before (`.. attribute:: name`)

**Comment by qhuang872** [comment]
Thank you so much for answering my question and clear my confusion.  --Qiwei

________________________________

**Comment by dangra** [comment]
superseded by #3011

**Pull request closed without merging**